### PR TITLE
Stock revamp

### PR DESCRIPTION
**Pull request by SagarKarwande**
## Why?
We need some different strategy to handle stock properly on Avia.

## This change addresses the need by:
`core`
--------
- Changed the product entity to allow tracking the inventory with following types:
  - None - We don't track the stock for a product.
  - Track by product - The stock is tracked by the product itself
  - Track by variant - The variant product have their own stock entries and the stock are tracked by the variants.
- Stock Item has new field to keep low stock level. Using this field we can alert the user for updating the stock.

`admin_app`
-----------------
- Product edit page have `Inventory` tab to set the tracking method for the product.
- Stock can be updated for each variant from `Option and SKU's` tab.

[delivers #162092265]

## Checklist

- [ ] I have read [CONTRIBUTING.md][contributing].
- [ ] My code follows the [style guidelines][contributing] of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [ ] I have updated the documentation wherever necessary.
- [ ] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

### CODE REVIEW

File: apps/admin_app/lib/admin_app_web/controllers/product_controller.ex [hunk 1]
```diff
@@ -100,6 +101,22 @@ defmodule AdminAppWeb.ProductController do
...
+      tracking_level = product_params["inventory_tracking"]
+
+      case tracking_level do
+        "product" ->
+          Inventory.set_inventory_tracking(product, tracking_level, params["stock"])
+
+        _ ->
+          Inventory.set_inventory_tracking(product, tracking_level)
+      end
+
```
Comment (pkrawat1) [inline apps/admin_app/lib/admin_app_web/controllers/product_controller.ex]: how about using pattern matching in the `set_inventory_tracking` function directly.

```elixir
with %ProductSchema{} = product <- ProductModel.get(params["product_id"])  do
  Inventory.set_inventory_tracking(product, tracking_level, params["stock"])
end
```

```elixir
def set_inventory_tracking(product, inventory_tracking, %{"inventory_tracking" => "product", "stock" => stock_params}) do
....
end

def set_inventory_tracking(product, inventory_tracking, %{ "stock" => stock_params}) do
...
end
```

File: apps/admin_app/lib/admin_app_web/views/product_view.ex [hunk 2]
```diff
@@ -36,6 +49,16 @@ defmodule AdminAppWeb.ProductView do
     length(product.taxon.variation_themes) > 0
   end
 
+  def is_variant_tracking_enabled(product) do
+    product.inventory_tracking == :variant
+  end
+
+  defp is_child_product(product) do
```
Comment (ashish173) [inline apps/admin_app/lib/admin_app_web/views/product_view.ex]: @SagarKarwande are you sure about placing this method in the `product_view`? can you justify the responsibility of the file and the this method?

File: apps/admin_app/lib/admin_app_web/views/product_view.ex [hunk 3]
```diff
@@ -36,6 +49,16 @@ defmodule AdminAppWeb.ProductView do
     length(product.taxon.variation_themes) > 0
   end
 
+  def is_variant_tracking_enabled(product) do
```
Comment (ashish173) [inline apps/admin_app/lib/admin_app_web/views/product_view.ex]: @SagarKarwande are you sure about placing this method in the `product_view`? can you justify the responsibility of the file and the this method?

File: apps/admin_app/lib/admin_app_web/views/product_view.ex [hunk 4]
```diff
@@ -16,6 +24,11 @@ defmodule AdminAppWeb.ProductView do
   @sort_field_keys ["rummage", "sort", "field"]
   @sort_order_keys ["rummage", "sort", "order"]
 
+  def active_stock_locaton() do
+    StockLocation.active()
+    |> Enum.map(fn x -> {x.name, x.id} end)
+  end
+
   def themes_options(product) do
```
Comment (ashish173) [inline apps/admin_app/lib/admin_app_web/views/product_view.ex]: @SagarKarwande are you sure about placing this method in the `product_view`? can you justify the responsibility of the file and the this method?

File: apps/admin_app/lib/admin_app_web/views/product_view.ex [hunk 5]
```diff
@@ -16,6 +24,11 @@ defmodule AdminAppWeb.ProductView do
   @sort_field_keys ["rummage", "sort", "field"]
   @sort_order_keys ["rummage", "sort", "order"]
 
+  def active_stock_locaton() do
```
Comment (ashish173) [inline apps/admin_app/lib/admin_app_web/views/product_view.ex]: @SagarKarwande I am not sure about this method either.

File: apps/admin_app/lib/admin_app_web/views/product_view.ex [hunk 6]
```diff
@@ -36,6 +49,16 @@ defmodule AdminAppWeb.ProductView do
...
+    product.inventory_tracking == :variant
+  end
+
+  defp is_child_product(product) do
+    query = from(c in Variation, where: c.child_product_id == ^product.id)
+    count = Repo.aggregate(query, :count, :id)
+    count > 0
+  end
+
   def has_variants(product) do
```
Comment (ashish173) [inline apps/admin_app/lib/admin_app_web/views/product_view.ex]: @SagarKarwande are you sure about placing this method in the `product_view`? can you justify the responsibility of the file and the this method?

File: apps/admin_app/lib/admin_app_web/views/product_view.ex [hunk 7]
```diff
@@ -218,4 +241,8 @@ defmodule AdminAppWeb.ProductView do
   defp generate_taxon_text([head | tail], acc) do
     generate_taxon_text(tail, acc <> "#{head.name} > ")
   end
+
+  def get_total_stock(variant) do
```
Comment (ashish173) [inline apps/admin_app/lib/admin_app_web/views/product_view.ex]: @SagarKarwande are you sure about placing this method in the `product_view`? can you justify the responsibility of the file and the this method?

File: apps/admin_app/lib/admin_app_web/views/product_view.ex (hunk 4 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/views/product_view.ex]: We write some function in `view` file that are specifically needed for templates.

`active_stock_locaton/0` is used to show select dropdown for active Stock Locations. Phoenix forms `select` need the data in one of the following forms:
```elixir
[[key: "Pune", value: 1 ], [key: "Mumbai", value: 2]] OR
["Pune": 1, "Mumbai": 2] OR
[{"Pune", 1}, {"Mumbai", 2}]
```
This is something that is specific that is needed by Phoenix. So I have used the `StockLocation` model to fetch the data and later transform it in to the way the `select` input needs it.

[Select input doc](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#select/4)
> I'll change the naming of variables in the method.

File: apps/admin_app/lib/admin_app_web/views/product_view.ex (hunk 6 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/views/product_view.ex]: This method was included in the merge commit, while I was resolving the conflicts after merging the `develop` branch. 

This method is already refactored in develop branch: [Develop link](https://github.com/aviacommerce/avia/blob/4809919dcb389880d626c8edc485a21280a69a39/apps/admin_app/lib/admin_app_web/views/product_view.ex#L27)

I'll remove the duplicate entry of this method.

File: apps/admin_app/lib/admin_app_web/views/product_view.ex (hunk 5 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/views/product_view.ex]: We write some function in `view` file that are specifically needed for templates.

`active_stock_locaton/0` is used to show select dropdown for active Stock Locations. Phoenix forms `select` need the data in one of the following forms:
```elixir
[[key: "Pune", value: 1 ], [key: "Mumbai", value: 2]] OR
["Pune": 1, "Mumbai": 2] OR
[{"Pune", 1}, {"Mumbai", 2}]
```
This is something that is specific that is needed by Phoenix. So I have used the `StockLocation` model to fetch the data and later transform it in to the way the `select` input needs it.

[Select input doc](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#select/4)
> I'll change the naming of variables in the method.

File: apps/admin_app/lib/admin_app_web/views/product_view.ex (hunk 4 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/views/product_view.ex]: We write some function in `view` file that are specifically needed for templates.

`themes_options/1` is used to show select dropdown for Variation themes. Phoenix forms `select` need the data in one of the following forms:
```elixir
[[key: "Color", value: 1 ], [key: "Color-size", value: 2]] OR
["Color": 1, "Color-Size": 2] OR
[{"Color", 1}, {"Color-Size", 2}]
```
This is something that is specific that is needed by Phoenix. So I have used the preloaded variation themes to fetch the data and later transform it into the way the `select` input needs it.

[Select input doc](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#select/4)

File: apps/admin_app/lib/admin_app_web/views/product_view.ex (hunk 2 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/views/product_view.ex]: This method was included in the merge commit, while I was resolving the conflicts after merging the `develop` branch. 

This method is already refactored in develop branch: [Develop link](https://github.com/aviacommerce/avia/blob/4809919dcb389880d626c8edc485a21280a69a39/apps/admin_app/lib/admin_app_web/views/product_view.ex#L27)

I'll remove the duplicate entry of this method.

File: apps/admin_app/lib/admin_app_web/views/product_view.ex (hunk 6 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/views/product_view.ex]: This method can be moved to `Product` model.

File: apps/admin_app/lib/admin_app_web/views/product_view.ex (hunk 3 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/views/product_view.ex]: I think we can move it to `Schema level` as it is related to Enum. Changing the comparison value would be quite easy later on.

File: apps/admin_app/lib/admin_app_web/controllers/product_controller.ex (hunk 1 above)
Comment (SagarKarwande) [reply apps/admin_app/lib/admin_app_web/controllers/product_controller.ex]: done

**Pull request merged**